column_reader.go

```diff
@@ -0,0 +1,872 @@
+		n += d
+	}
+	return n, err
+}
+
+func (r *booleanColumnReader) ReadValues(values []Value) (n int, err error) {
+	if cap(r.buffer) == 0 {
+		r.buffer = make([]bool, 0, atLeastOne(r.bufferSize))
+	}
+
+	for {
+		for r.offset < len(r.buffer) && n < len(values) {
+			values[n] = makeValueBoolean(r.buffer[r.offset])
+			values[n].columnIndex = r.columnIndex
+			r.offset++
+			n++
+		}
+
+		if n == len(values) {
+			return n, nil
+		}
+		if r.decoder == nil {
+			return n, io.EOF
+		}
+
+		buffer := r.buffer[:cap(r.buffer)]
+		d, err := r.decoder.DecodeBoolean(buffer)
+		if d == 0 {
+			return n, err
+		}
+
+		r.buffer = buffer[:d]
+		r.offset = 0
+	}
+}
+
+func (r *booleanColumnReader) Reset(decoder encoding.Decoder) {
+	r.decoder = decoder
+	r.buffer = r.buffer[:0]
+	r.offset = 0
+}
+
+type int32ColumnReader struct {
+	typ         Type
+	decoder     encoding.Decoder
+	buffer      []int32
+	offset      int
+	bufferSize  int
+	columnIndex int16
+}
+
+func newInt32ColumnReader(typ Type, columnIndex int16, bufferSize int) *int32ColumnReader {
+	return &int32ColumnReader{
+		typ:         typ,
+		bufferSize:  bufferSize,
+		columnIndex: ^columnIndex,
+	}
+}
+
+func (r *int32ColumnReader) Type() Type { return r.typ }
+
+func (r *int32ColumnReader) Column() int { return int(^r.columnIndex) }
+
+func (r *int32ColumnReader) ReadInt32s(values []int32) (n int, err error) {
+	if r.offset < len(r.buffer) {
+		n = copy(values, r.buffer[r.offset:])
+		r.offset += n
+		values = values[n:]
+	}
+	if r.decoder == nil {
+		err = io.EOF
+	} else {
+		var d int
+		d, err = r.decoder.DecodeInt32(values)
+		n += d
+	}
+	return n, err
+}
+
+func (r *int32ColumnReader) ReadValues(values []Value) (n int, err error) {
+	if cap(r.buffer) == 0 {
+		r.buffer = make([]int32, 0, atLeastOne(r.bufferSize))
+	}
+
+	for {
+		for r.offset < len(r.buffer) && n < len(values) {
+			values[n] = makeValueInt32(r.buffer[r.offset])
+			values[n].columnIndex = r.columnIndex
+			r.offset++
+			n++
+		}
+
+		if n == len(values) {
+			return n, nil
+		}
+		if r.decoder == nil {
+			return n, io.EOF
+		}
+
+		buffer := r.buffer[:cap(r.buffer)]
+		d, err := r.decoder.DecodeInt32(buffer)
+		if d == 0 {
+			return n, err
+		}
+
+		r.buffer = buffer[:d]
+		r.offset = 0
+	}
+}
+
+func (r *int32ColumnReader) Reset(decoder encoding.Decoder) {
+	r.decoder = decoder
+	r.buffer = r.buffer[:0]
+	r.offset = 0
+}
+
+type int64ColumnReader struct {
+	typ         Type
+	decoder     encoding.Decoder
+	buffer      []int64
+	offset      int
+	bufferSize  int
+	columnIndex int16
+}
+
+func newInt64ColumnReader(typ Type, columnIndex int16, bufferSize int) *int64ColumnReader {
+	return &int64ColumnReader{
+		typ:         typ,
+		bufferSize:  bufferSize,
+		columnIndex: ^columnIndex,
+	}
+}
+
+func (r *int64ColumnReader) Type() Type { return r.typ }
+
+func (r *int64ColumnReader) Column() int { return int(^r.columnIndex) }
+
+func (r *int64ColumnReader) ReadInt64s(values []int64) (n int, err error) {
```

Comment:
I believe this is meant so we can more easily convert parquet to arrow. In that case I think it would be very helpful, if the functions that return the native types such as this one, read a validity boolean mask as well so the read result can be passed directly to the respective type's `AppendValues` function. [`Int64Builder.AppendValues`](https://pkg.go.dev/github.com/apache/arrow/go/arrow@v0.0.0-20211112161151-bc219186db40/array#Int64Builder.AppendValues) in this case.

Reply:
Arrow integration is one of the intentions here indeed. 

I've been thinking about how to introduce these bitmaps of null values, I'm currently edging towards having a second set of interfaces for optional columns, with methods that support loading both the values and bitmaps, for example:

```go
type OptionalInt32Reader interface {
  ReadOptionalInt32s(nulls bitmap, values []int32) (int, error)
}
```

Separating the methods can give stronger type checking guarantees and avoids pushing the complexity of dealing with the bitmaps in code that doesn't need it... on the other hand, it's a lot more code to manage in the library. 

Let me know if you have suggestions

Reply:
I like the additional set of interfaces!